### PR TITLE
add asciidoc render of Program Guide

### PR DESCRIPTION
**Pull request by clevinson**
#### TL;DR

The following PR sets up a mechanism for maintaining the program guide in a lightweight markup format (Asciidoc), and having scripts that can generate HTML or PDF versions of the program guide with the appropriate [decimal outline format](https://writingcenter.tamu.edu/Grads/Writing-Speaking-Guides/Alphabetical-List-of-Guides/Brainstorming-Planning/Decimal-Outlines).

Previews:
- [plaintext view of program-guide.adoc file](https://github.com/regen-network/regen-registry-program-guide/blob/clev/markdown/program-guide.adoc?plain=1)
- [rendered HTML view](https://deploy-preview-1--lustrous-heliotrope-bbde81.netlify.app/)
- [rendered PDF view](https://github.com/regen-network/regen-registry-program-guide/blob/clev/markdown/pdf/program-guide.pdf)
- [github native rendering of .adoc](https://github.com/regen-network/regen-registry-program-guide/blob/clev/markdown/program-guide.adoc)

## Overview

After testing out a few different markup languages (md, rst, latex, asciidoc), I found that asciidoc has the best support for automatic [decimal outline format](https://writingcenter.tamu.edu/Grads/Writing-Speaking-Guides/Alphabetical-List-of-Guides/Brainstorming-Planning/Decimal-Outlines), while still being an easily readable lightweight markup language. Latex is probably the more fully-featured of the options I explored, but I found AsciiDoc easier to manage and work with as it has wider support for preview-renderers (both in code editors, and in GitHub).

For a lightweight overview of how I'm using AsciiDoc in the context of the Program Guide see [the readme](https://github.com/regen-network/regen-registry-program-guide/blob/clev/markdown/README.md) of this PR's branch.

There are a few nuances in our current usage of Decimal Outline formatting that make it a bit challenge even in Asciidoc which I'd like to talk through here:
1. Decimal outline notation is used both for section headings ("5." or "5.1"), as well as for raw outline content ("5.1.1." or "5.1.2.")
2. Sometimes we have un-numbered subsection headings interspersed between enumerated items in an outline

For a visual example of what I'm talking about, see this snippet from [Section 5.2](https://docs.google.com/document/d/1emUjpRf1KiDR42WhMWc_zdN9sGZhQXfLsqWwWXyfMXM/edit#heading=h.lhbb2owsk2kz) below:
Screenshot 2023-08-04 at 12 50 24 PM

AsciiDoc has native support for decimal notation in section headings, and can do up to 5 levels of sub-headings (e.g. 5.1.1.2.3.).  But using ordered lists for actual content does not inherit the decimal outline notation from the relevant subheadings.

Here's my suggestions for how we work around these issues:

### Part 1: Leverage Asciidoc's native support for [Section Numbers](https://docs.asciidoctor.org/asciidoc/latest/sections/numbers)

Manually editing decimal outline numerals is something I'm trying to avoid, so I'm hoping we can find a solution that leverages auto-generated numbers where possible. Thankfully, asciidoc has native support for [section numbers](https://docs.asciidoctor.org/asciidoc/latest/sections/numbers).

If we have an ordered list within the scope of as section, the standard rendering of it would be as follows:

```
=== New Credit Classes and Methodologies
. All new Credit Classes.....
. Regen Registry may grandfather in new Methodologies....
. Regen Registry must document the entire Credit Class ....
```
would render:
> ### 5.1. New Credit Classes and Methodologies
> 1. All new Credit Classes.....
> 2. Regen Registry may grandfather in new Methodologies....
> 3. Regen Registry must document the entire Credit Class ....

Without any custom stylesheets.  This means that it would appear as above in GitHub.  However, I think we could introduce some tooling that would actually generate PDF and/or HTML files with the appropriate rendering.  This would require adding [custom stylesheets](https://docs.asciidoctor.org/asciidoctor/latest/html-backend/custom-stylesheet/) which can be incorporated during HTML generation, and also when [converting an .adoc file to pdf](https://github.com/ggrossetie/asciidoctor-web-pdf).

The stylesheet would simply need to override all ordered lists displays and prepend them with the numbers from the a "Section Heading" parent div's number.  I'm not 100% how this would work, but I'm hopeful that with some css magic we can get it sorted.  Worst case we could add a javascript post-processing step to do this, and have PDFs generated from the post-javascript processed HTML.

### Part 2: Get decimal outlines working properly with interspersed sub-sub-headings

As you can see in the image above, subsection headings such as "Proposed Modifications to Approved Credit Classes and Methodologies" and "Credit Class and Methodology Update Requirements for Existing Projects" work quite nicely in google docs. To make this work in asciiDoc however, we have 2 options:

#### Option 2a) Turn sub-sub-headings into "decimal outline" numbered sections

By setting `:sectnumlevels: 3` in asciiDoc, we turn each of these sub-sub headings into their own decimal outline numbered header.  Doing so means that it's fine to reset the index of an ordered list after each new sub-sub-heading, since the decimal outline element is incrementing with each sub-sub-heading as well.

To illustrate by example, this would mean:

```adoc
:sectnumlevels: 3

=== Modifications to an Approved Credit Class or Methodology

==== Proposed Modifications to Approved Credit Classes and Methodologies

. Credit Class Admins may propose ....
. Regen Registry, currently managed by RND, will review the extent ....
. If changes to Credit Class or Methodology are approved by the Regen Registry,....

==== Credit Class and Methodology Update Requirements for Existing Projects

. If a new version of a Credit Class ...
. Project Proponents are required to implement...

```

Would render (in GitHub, and under normal styling) as:

> ### 5.2. Modifications to an Approved Credit Class or Methodology
>
> #### 5.2.1. Proposed Modifications to Approved Credit Classes and Methodologies
>
> 1. Credit Class Admins may propose ....
> 2. Regen Registry, currently managed by RND, will review the extent ....
> 3. If changes to Credit Class or Methodology are approved by the Regen Registry,....
> 
> #### 5.2.2. Credit Class and Methodology Update Requirements for Existing Projects
> 
> 1. If a new version of a Credit Class ...
> 2. Project Proponents are required to implement...

This way, the same post-processed (CSS or JS) described in **Part A** could be leveraged here, and all would be fine and dandy (the end result in a PDF would be as follows:

> ### 5.2. Modifications to an Approved Credit Class or Methodology
>
> #### 5.2.1. Proposed Modifications to Approved Credit Classes and Methodologies
>
> 5.2.1.1. Credit Class Admins may propose ....
> 5.2.1.2. Regen Registry, currently managed by RND, will review the extent ....
> 5.2.1.3. If changes to Credit Class or Methodology are approved by the Regen Registry,....
> 
> #### 5.2.2. Credit Class and Methodology Update Requirements for Existing Projects
> 
> 5.2.2.1. If a new version of a Credit Class ...
> 5.2.2.2. Project Proponents are required to implement...

#### Option 2b) Explicitly reset the outline start index when interspersing a list with new sub-sub-headings

In this case, we would have `:sectnumlevels: 2`, and would need to manually override the start of an ordered list for the formatting to work correctly.

Example:

```adoc
:sectnumlevels: 2

=== Modifications to an Approved Credit Class or Methodology

==== Proposed Modifications to Approved Credit Classes and Methodologies

. Credit Class Admins may propose ....
. Regen Registry, currently managed by RND, will review the extent ....
. If changes to Credit Class or Methodology are approved by the Regen Registry,....

==== Credit Class and Methodology Update Requirements for Existing Projects

[start=4]
. If a new version of a Credit Class ...
. Project Proponents are required to implement...

```

The default Github rendering would then be:

> ### 5.2. Modifications to an Approved Credit Class or Methodology
>
> #### Proposed Modifications to Approved Credit Classes and Methodologies
>
> 1. Credit Class Admins may propose ....
> 2. Regen Registry, currently managed by RND, will review the extent ....
> 3. If changes to Credit Class or Methodology are approved by the Regen Registry,....
> 
> #### Credit Class and Methodology Update Requirements for Existing Projects
> 
> 4. If a new version of a Credit Class ...
> 5. Project Proponents are required to implement...

And then post-processed rendering would hopefully be:

> ### 5.2. Modifications to an Approved Credit Class or Methodology
>
> #### Proposed Modifications to Approved Credit Classes and Methodologies
>
> 5.2.1. Credit Class Admins may propose ....
> 5.2.2. Regen Registry, currently managed by RND, will review the extent ....
> 5.2.3. If changes to Credit Class or Methodology are approved by the Regen Registry,....
> 
> #### Credit Class and Methodology Update Requirements for Existing Projects
> 
> 5.2.4. If a new version of a Credit Class ...
> 5.2.5. Project Proponents are required to implement...

### Conclusion
My preference is to focus first on **Part 1**, and as a next step see what kind of CSS / js post-processing we can do to get decimal outline notation functioning for all ordered lists as expected.

Once we do that, we can look at **Part 2**. My current preference is for **Option 2a**. As I think it will be cleaner from a maintenance perspective.  There is a chance that we could find a third **Option 2c** solution which involves no changes, and just leverages further post-processing to get things formatted exactly like in google docs, but I don't know how to evaluate the complexity of that option without getting through **Part 1** first.

Curious to hear others' thoughts & opinions on this!

### CODE REVIEW

File: stylesheets/styles.css [hunk 1]
```diff
@@ -0,0 +1,30 @@
+/* Include default AsciiDoc styles here... */
+@import "https://fonts.googleapis.com/css?family=Open+Sans:300,300italic,400,400italic,600,600italic%7CNoto+Serif:400,400italic,700,700italic%7CDroid+Sans+Mono:400,700";
+@import "https://cdn.jsdelivr.net/gh/asciidoctor/asciidoctor@2.0/data/stylesheets/asciidoctor-default.css";
```
Comment (clevinson) [inline stylesheets/styles.css]: Should probably include this default stylesheet explicitly instead of depending on asciidoc's CDN if we are going to keep it.

**Comment by clevinson** [comment]
OK so I've got a minimal implementation working now.

You can see the [HTML Preview](https://deploy-preview-1--lustrous-heliotrope-bbde81.netlify.app/) to see how this renders, as well as the [PDF preview](https://github.com/regen-network/regen-registry-program-guide/blob/clev/markdown/pdf/program-guide.pdf) here.

This implementation is going w/ **Option 2a** currently.  I'd like to pause work on this for now, and get feedback from @S4mmyb or other registry team members about the current web-based rendering of the program guide before proceeding further.

Once I get feedback on any outstanding issues w/ the current approach, I can try to resolve them and do some general clean up on the design.

File: src/decimal-outline-pdf.js [hunk 2]
```diff
@@ -0,0 +1,53 @@
+const asciidoctor = require("asciidoctor")();
```
Comment (clevinson) [inline src/decimal-outline-pdf.js]: These two custom converters should be de-duplicated, and the bulk of the "decimal-outline" logic moved to its own single .js file.

**Comment by wgwz** [comment]
@clevinson the html view looks great!

**Comment by wgwz** [comment]
@clevinson are there any tools available that could make the editing experience better for the adoc format? i personally like the simplicity of editing a plain text document, but i know for others they might want kind of like the hackmd style split screen view or a more visual tool. just wondering if you looked into that at all yet, this looks great though!

**Comment by wgwz** [comment]
@clevinson curious what happened here, this was looking good!
